modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/internal/shmem/ShmemAcceptWorker.java

```diff
@@ -0,0 +1,141 @@
+import org.apache.ignite.internal.util.nio.GridNioServerListener;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.internal.util.worker.GridWorker;
+import org.apache.ignite.plugin.extensions.communication.Message;
+import org.apache.ignite.plugin.extensions.communication.MessageFactory;
+import org.apache.ignite.spi.communication.tcp.TcpCommunicationMetricsListener;
+import org.apache.ignite.thread.IgniteThread;
+
+/**
+ * This worker takes responsibility to shut the server down when stopping, No other thread shall stop passed server.
+ */
+public class ShmemAcceptWorker extends GridWorker {
+    /** Worker name. */
+    private static final String WORKER_NAME = "shmem-communication-acceptor";
+
+    /** Shared memory workers. */
+    private final Collection<ShmemWorker> shmemWorkers = new ConcurrentLinkedDeque<>();
+
+    /** Server. */
+    private final IpcSharedMemoryServerEndpoint srv;
+
+    /** Server listener. */
+    private final GridNioServerListener<Message> srvLsnr;
+
+    /** Statistics. */
+    private final TcpCommunicationMetricsListener metricsLsnr;
+
+    /** Logger. */
+    private final IgniteLogger log;
+
+    /** Message factory. */
+    private final MessageFactory msgFactory;
+
+    /** Writer factory. */
+    private final GridNioMessageWriterFactory writerFactory;
+
+    /** Reader factory. */
+    private final GridNioMessageReaderFactory readerFactory;
+
+    /** Tracing. */
+    private final Tracing tracing;
+
+    /**
+     * @param igniteInstanceName Ignite instance name.
+     * @param srvLsnr Server listener.
+     * @param srv Server.
+     * @param metricsLsnr Metrics listener.
+     * @param log Logger.
+     * @param msgFactory Message factory.
+     * @param writerFactory Writer factory.
+     * @param readerFactory Reader factory.
+     * @param tracing Tracing.
+     */
+    public ShmemAcceptWorker(
+        String igniteInstanceName,
+        GridNioServerListener<Message> srvLsnr,
+        IpcSharedMemoryServerEndpoint srv,
+        TcpCommunicationMetricsListener metricsLsnr,
+        IgniteLogger log,
+        MessageFactory msgFactory,
+        GridNioMessageWriterFactory writerFactory,
+        GridNioMessageReaderFactory readerFactory,
+        Tracing tracing
+    ) {
+        super(igniteInstanceName, WORKER_NAME, log);
+
+        this.msgFactory = msgFactory;
+        this.writerFactory = writerFactory;
+        this.readerFactory = readerFactory;
+        this.tracing = tracing;
+        this.srv = srv;
+        this.srvLsnr = srvLsnr;
+        this.metricsLsnr = metricsLsnr;
+        this.log = log;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void body() throws InterruptedException {
+        try {
+            while (!Thread.interrupted()) {
+                ShmemWorker e = new ShmemWorker(
+                    igniteInstanceName(),
+                    log,
+                    tracing,
+                    srv.accept(),
```

Comment:
Let's create a local variable for this, as it is not obvious that a thread blocks here.
Now it looks like worker is spinning in a loop and producing infinite number of threads.

Also, would it be reasonable to cache unused ShmemWorkers in some fixed-length buffer\queue?